### PR TITLE
[feat] Add restore to base loop

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds restore to base loop and is_restarting boolean.

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
- what's the difference between this and load_state_dict?
- why do we need the is_restarting boolean?

File: tests/loops/test_loops.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    data = range(10)
+    loop = Simple(data)
+    try:
+        loop.run()
+    except CustomExpection:
+        state_dict = loop.state_dict()
+
+    loop = Simple(data)
+    loop.load_state_dict(state_dict)
+    loop.is_restarting = True
```
Comment (yifuwang) [inline tests/loops/test_loops.py:72]: noob question: does `load_state_dict` imply `is_restarting` and vice versa? Just wondering if the purpose of `is_restarting` and `restore` can be served by `load_state_dict` alone.

**Comment by tchaton** [comment]
> * what's the difference between this and load_state_dict?
> * why do we need the is_restarting boolean?

On restart, loading the state dict isn't enough, you would need to do multiple things depending on your loop such as fast forwarding samplers, etc..

Best,
T.C

File: tests/loops/test_loops.py (hunk 1 above)
Comment (tchaton) [reply tests/loops/test_loops.py:72]: Yes, great question ! @carmocca @awaelchli.
Should `load_state_dict` automatically set `is_restarting` property to True ?

File: pytorch_lightning/loops/base.py [hunk 2]
```diff
@@ -33,6 +33,12 @@ class Loop(ABC):
 
     .. codeblock:: python
 
+        if is_restarting:
+            restore()
+            is_restarting = False
+        else:
+            reset()
+
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:41]: How many details do we want to include in this docs overview? 
Need to be careful what we want to show, we don't want to start duplicating the implementation into the docs.

The raw loop structure below I think is already good as it is to just understand how the flow goes and where the core component is.

File: tests/loops/test_loops.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+from pytorch_lightning.loops.base import Loop
+
+
+def test_loop_restore():
+
+    class CustomExpection(Exception):
+        pass
+
+    class Simple(Loop):
```
Comment (awaelchli) [inline tests/loops/test_loops.py:25]: This brings me to the idea of a "BoringLoop", which could be used in the future for unit testing our custom loops. 
I will look into this as part of my api exploration for customization <3

File: tests/loops/test_loops.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        def done(self) -> bool:
+            return self.iteration_count > len(self.dataset)
+
+        def reset(self) -> None:
+            self.iter_dataset = iter(self.dataset)
+            self.outputs = []
+
+        def advance(self) -> None:
+            value = next(self.iter_dataset)
+            print(value, self.iteration_count)
```
Comment (awaelchli) [inline tests/loops/test_loops.py:47]: leftover print statement

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/base.py [hunk 5]
```diff
@@ -46,6 +52,15 @@ class Loop(ABC):
     def __init__(self) -> None:
         self.iteration_count: int = 0
         self.trainer: Optional['pl.Trainer'] = None
+        self._is_restarting = False
+
+    @property
+    def is_restarting(self) -> bool:
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:58]: Should the name be `restarting` to be consistent with `done`? It's not `is_done`

File: tests/loops/test_loops.py (hunk 1 above)
Comment (carmocca) [reply tests/loops/test_loops.py:72]: Mmm that is potentially dangerous, we could wait and see if every case of `load_state_dict` will need to set `is_restarting` and if it's worth removing the extra line in that case

File: tests/loops/test_loops.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        def load_state_dict(self, state_dict: Dict) -> None:
+            self.iteration_count = state_dict["iteration_count"]
+            self.outputs = state_dict["outputs"]
+
+    state_dict = {}
+
+    data = range(10)
+    loop = Simple(data)
+    try:
+        loop.run()
```
Comment (carmocca) [inline tests/loops/test_loops.py:66]:
```suggestion
    data = range(10)
    loop = Simple(data)
    try:
        loop.run()
        state_dict = {}
```

File: pytorch_lightning/loops/base.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:41]: Removed !

File: tests/loops/test_loops.py (hunk 4 above)
Comment (tchaton) [reply tests/loops/test_loops.py:47]: Removed !

File: pytorch_lightning/loops/base.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:58]: Done !

File: tests/loops/test_loops.py (hunk 6 above)
Comment (tchaton) [reply tests/loops/test_loops.py:66]: Done !

**Review by kaushikb11 — APPROVED** [review]
LGTM!

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/loops/base.py [hunk 7]
```diff
@@ -46,6 +46,15 @@ class Loop(ABC):
...
         self.trainer: Optional['pl.Trainer'] = None
+        self._restarting = False
+
+    @property
+    def restarting(self) -> bool:
+        return self._restarting
+
+    @restarting.setter
+    def restarting(self, restarting: bool) -> None:
+        self._restarting = restarting
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:57]: Is there a reason why you added a getter/setter considering they don't do anything custom?

Comment (tchaton) [reply pytorch_lightning/loops/base.py:57]: Not a strong need for them right now. Can be removed later on.
